mapper/pojo/src/main/java/org/hibernate/search/mapper/pojo/work/impl/PojoWorkPlanImpl.java

```diff
@@ -108,14 +110,24 @@ public void purge(Class<?> clazz, Object providedId) {
 
 	@Override
 	public void prepare() {
-		for ( PojoContainedTypeWorkPlan<?> delegate : containedTypeDelegates.values() ) {
-			delegate.resolveDirty( this::updateBecauseOfContained );
+		if ( isPreparing ) {
```

Comment:
This looks... odd. Can you tell me how to reproduce the issue you're working around? I'm concerned we may need to do more than just ignoring the call...

Reply:
You're right. Actually I was also skeptical about that.

Without the check, we get a `StackOverflow` running `AutomaticIndexingBridgeExplicitReindexingFunctionalIT.test`.

In the stack trace we can see an infinite loop of `HibernateSearchEventListener.onAutoFlush`. Where each of ones invokes a prepare on the same `PojoWorkPlanImpl` instance.

Reply:
So... I investigated and it turns out that, in most cases, this recursion shouldn't happen.

In this test, the recursion only happens because we execute a search query inside a bridge, and that search query triggers an automatic flush. But, since we flush just before we execute the bridges, there shouldn't be anything to flush.

So, I added a few commits:

1. Throw an exception when such recursion happens, because it's definitely not normal.
2. Disable automatic flushes while processing entities for indexing.
3. I'll get back to the third one.

Unfortunately, this is not enough. It turns out the `AutoFlushEvent` is fired even when no flush happens: it is actually used to tell the `DefaultAutoFlushEventListener` "please check whether an automatic flush is required, and if so, do it".

This means that our current behavior of considering a flush occurred every time we get an `AutoFlushEvent` is wrong: it just means a flush was requested, but it may not have occurred.

So we need to check whether the flush actually occurred. I added a third commit that tries to leverage `AutoFlushEvent#isFlushRequired`. Unfortunately this method always return `false`, because of what I think is a bug in ORM.

I'll open a ticket for ORM, but it likely will take a few weeks before the fix is released...

In the meantime, maybe you could explore an alternative solution: `org.hibernate.SessionEventListener#partialFlushEnd`. This method is called after an automatic flush is actually performed, and only if it is actually performed. Ignore the "partial" in the name, the flush is not partial at all.
So you could make `HibernateSearchEventListener` extend `BaseSessionEventListener` and move the implementation of the auto flush listener to `partialFlushEnd`.
The problem will be how to register the session listener, and unfortunately I don't have an answer for that... Maybe you can investigate?

Reply:
Created https://hibernate.atlassian.net/browse/HHH-13592

Reply:
Thanks @yrodiere. 

From my investigation trying to register the listener at `SessionFactory` level there is an hibernate property `hibernate.session.events.auto` to set the fully qualified class name implementing the `SessionEventListener` interface.
It seems that there is no service to invoke to register it, like for other events.
Moreover, it is supposed to exist only one listener. So if we registered our listener, the user couldn't do the same for his/hers.

Trying to register the listener at `Session` level, that it means do it for each session we create...
It seems to be add later session listeners. I know it's no very nice, but I've not found a better solution.

I'm following the latter. Let's see if it will bring us to a solution.

Reply:
@yrodiere it seems that switching from `onAutoFlush` to `partialFlushEnd` produces the same results. Looking at Hibernate ORM source code it seems that `autoflush` always triggers a `partialFlushEnd` event.

What would I do? For now I would make idempotent the `PojoWorkPlan` instance again, waiting for the completion of HHH-13592. Opening a follow-up of HHH-13592 on Search side.

On the other hand, any other ideas are more than welcome.